Bug in Astro with Headers

### Is this suited for github?

- [x] Yes, this is suited for github

### Reproduction

1. Configure an Astro project with output server a server island that requires checking the session
2. Use betther-auth 1.6.15 or above
3. Start a dev server and access it

You will encounter an error:

```bash
17:27:55 [ERROR] APIError: Headers is required
17:27:55 [ERROR] Headers is required
```

### Current vs. Expected behavior

Prior this versions the web launches normally, it loads and then the server island loads when it finishes its rendering and checkin session.

After 1.6.15 (included) it shows a `APIError: Headers is required`.

<img width="448" height="198" alt="Image" src="https://github.com/user-attachments/assets/d6aa47f4-14f0-4951-bb98-c1e2af538c7c" />

<img width="1287" height="451" alt="Image" src="https://github.com/user-attachments/assets/d358283e-6c62-471b-903e-6baa527d7d0b" />

*If all server islads with sessions are removed, it works fine.

### What version of Better Auth are you using?

1.6.15 or above

### System info

```bash
{
  "system": {
    "platform": "win32",
    "arch": "x64",
    "version": "Windows 11 Home",
    "release": "10.0.26200",
    "cpuCount": 28,
    "cpuModel": "Intel(R) Core(TM) i7-14700K",
    "totalMemory": "31.72 GB",
    "freeMemory": "5.90 GB"
  },
  "node": {
    "version": "v26.3.0",
    "env": "development"
  },
  "packageManager": {
    "name": "npm",
    "version": "11.16.0"
  },
  "frameworks": [
    {
      "name": "astro",
      "version": "6.4.6"
    }
  ],
  "databases": [
    {
      "name": "mysql2",
      "version": "3.22.5"
    },
    {
      "name": "drizzle",
      "version": "0.45.2"
    }
  ],
  "betterAuth": {
    "version": "1.6.15",
    "config": {
      "baseURL": "http://localhost:4321",
      "secret": "[REDACTED]",
      "trustedOrigins": [
        "https://lioncommunity.es",
        "http://localhost:4321"
      ],
      "advanced": {
        "defaultCookieAttributes": {
          "sameSite": "lax"
        }
      },
      "user": {
        "additionalFields": {
          "discordId": {
            "type": "string",
            "required": false,
            "input": false
          }
        }
      },
      "session": {
        "cookieCache": {
          "enabled": true,
          "maxAge": 34560000,
          "strategy": "jwe",
          "refreshCache": true
        }
      },
      "account": {
        "storeStateStrategy": "cookie",
        "storeAccountCookie": true
      },
      "socialProviders": {
        "discord": {
          "clientId": "[REDACTED]",
          "clientSecret": "[REDACTED]",
          "scope": [
            "identify",
            "email"
          ]
        }
      }
    }
  }
}
```

### Which area(s) are affected? (Select all that apply)

Backend

### Auth config (if applicable)

```typescript

```

### Additional context

It's both reproducible online and locally.

--- a/.changeset/quiet-astro-sessions.md
+++ b/.changeset/quiet-astro-sessions.md
@@ -1,0 +1,5 @@
+---
+"better-auth": patch
+---
+
+Allow direct `auth.api.getSession()` calls without request headers to return `null` instead of throwing when no session cookies are available, including custom-session responses.

--- a/docs/content/docs/basic-usage.mdx
+++ b/docs/content/docs/basic-usage.mdx
@@ -292,7 +292,7 @@ You can also use it with client-side data-fetching libraries like [TanStack Quer
 
 ### Server Side
 
-The server provides a `session` object that you can use to access the session data. It requires request headers object to be passed to the `getSession` method.
+The server provides a `session` object that you can use to access the session data. Pass the current request headers to read the session from cookies. If no request headers are available, `auth.api.getSession` returns `null`.
 
 **Example: Using some popular frameworks**
 
@@ -303,7 +303,7 @@ The server provides a `session` object that you can use to access the session da
     import { headers } from "next/headers";
 
     const session = await auth.api.getSession({
-        headers: await headers() // you need to pass the headers object.
+        headers: await headers() // pass the headers object from the current request
     })
     ```
   </Tab>

--- a/docs/content/docs/integrations/astro.mdx
+++ b/docs/content/docs/integrations/astro.mdx
@@ -118,6 +118,8 @@ export const onRequest = defineMiddleware(async (context, next) => {
 });
 ```
 
+For server islands, prefer reading the session from `Astro.locals` after middleware has populated it, or pass `Astro.request.headers` when calling `auth.api.getSession` directly. A direct `auth.api.getSession` call without a request source returns `null`.
+
 ### Getting session on the server inside `.astro` file
 
 You can use `Astro.locals` to check if the user has session and get the user data from the server side. Here is an example of how you can get the session inside an `.astro` file:

--- a/docs/content/docs/reference/faq.mdx
+++ b/docs/content/docs/reference/faq.mdx
@@ -23,7 +23,7 @@ This page contains frequently asked questions, common issues, and other helpful 
   </Accordion>
 
   <Accordion id="getsession-not-working" title="getSession not working">
-    If you try to call `authClient.getSession` on a server environment (e.g, a Next.js server component), it doesn't work since it can't access the cookies. You can use the `auth.api.getSession` instead and pass the request headers to it.
+    If you try to call `authClient.getSession` on a server environment (e.g, a Next.js server component), it doesn't work since it can't access the cookies. You can use `auth.api.getSession` instead and pass the request headers when you want it to read the current request cookies. Without a request source, `auth.api.getSession` returns `null`.
 
     ```tsx title="server.tsx"
     import { auth } from "./auth";

--- a/packages/better-auth/src/api/routes/session-api.test.ts
+++ b/packages/better-auth/src/api/routes/session-api.test.ts
@@ -72,6 +72,13 @@ describe("session", async () => {
 		expect(response.data).toBeNull();
 	});
 
+	/**
+	 * @see https://github.com/better-auth/better-auth/issues/10044
+	 */
+	it("should return null for direct session checks without a request source", async () => {
+		await expect(auth.api.getSession()).resolves.toBeNull();
+	});
+
 	it("should require a fresh session based on session creation time", async () => {
 		vi.useFakeTimers();
 		const now = new Date("2026-01-01T00:00:00.000Z");

--- a/packages/better-auth/src/api/routes/session.ts
+++ b/packages/better-auth/src/api/routes/session.ts
@@ -36,7 +36,6 @@ export const getSession = <Option extends BetterAuthOptions>() =>
 			method: ["GET", "POST"],
 			operationId: "getSession",
 			query: getSessionQuerySchema,
-			requireHeaders: true,
 			metadata: {
 				openapi: {
 					operationId: "getSession",

--- a/packages/better-auth/src/api/to-auth-endpoints.test.ts
+++ b/packages/better-auth/src/api/to-auth-endpoints.test.ts
@@ -146,6 +146,22 @@ describe("before hook", async () => {
 			expect(res).toMatchObject({ key: "value", name: "headers" });
 		});
 
+		it("should keep requireHeaders endpoints strict without headers", async () => {
+			const endpoints = {
+				headers: createAuthEndpoint(
+					"/strict-headers",
+					{
+						method: "GET",
+						requireHeaders: true,
+					},
+					async (c) => Object.fromEntries(c.headers.entries()),
+				),
+			};
+			const authEndpoints = toAuthEndpoints(endpoints, init({}));
+
+			await expect(authEndpoints.headers()).rejects.toSatisfy(isAPIError);
+		});
+
 		it("should replace existing array when hook provides another array", async () => {
 			const endpoint = {
 				body: createAuthEndpoint(

--- a/packages/better-auth/src/plugins/custom-session/custom-session.test.ts
+++ b/packages/better-auth/src/plugins/custom-session/custom-session.test.ts
@@ -65,6 +65,13 @@ describe("Custom Session Plugin Tests", async () => {
 		expect(session?.newData).toEqual({ message: "Hello, World!" });
 	});
 
+	/**
+	 * @see https://github.com/better-auth/better-auth/issues/10044
+	 */
+	it("should return null for direct custom session checks without a request source", async () => {
+		await expect(auth.api.getSession()).resolves.toBeNull();
+	});
+
 	it("should return set cookie headers as separate entries", async () => {
 		const { headers } = await signInWithTestUser();
 		await client.getSession({

--- a/packages/better-auth/src/plugins/custom-session/index.ts
+++ b/packages/better-auth/src/plugins/custom-session/index.ts
@@ -95,7 +95,6 @@ export const customSession = <
 							},
 						},
 					},
-					requireHeaders: true,
 				},
 				async (ctx): Promise<Returns | null> => {
 					const session = await getSession()({

